Sources/NIOHTTPClient/HTTPHandler.swift

```diff
@@ -300,33 +319,44 @@ internal class TaskHandler<T: HTTPClientResponseDelegate>: ChannelInboundHandler
 
         context.write(wrapOutboundOut(.head(head)), promise: nil)
 
-        if let body = request.body {
-            let part: HTTPClientRequestPart
-            switch body {
-            case .byteBuffer(let buffer):
-                part = HTTPClientRequestPart.body(.byteBuffer(buffer))
-            case .data(let data):
-                var buffer = context.channel.allocator.buffer(capacity: data.count)
-                buffer.writeBytes(data)
-                part = HTTPClientRequestPart.body(.byteBuffer(buffer))
-            case .string(let string):
-                var buffer = context.channel.allocator.buffer(capacity: string.utf8.count)
-                buffer.writeString(string)
-                part = HTTPClientRequestPart.body(.byteBuffer(buffer))
-            }
+        self.writeBody(request: request, context: context).whenComplete { result in
+            switch result {
+            case .success:
+                context.write(self.wrapOutboundOut(.end(nil)), promise: promise)
+                context.flush()
 
-            context.write(wrapOutboundOut(part), promise: nil)
-        }
+                self.state = .sent
+                self.delegate.didTransmitRequestBody(task: self.task)
 
-        context.write(wrapOutboundOut(.end(nil)), promise: promise)
-        context.flush()
+                let channel = context.channel
+                self.promise.futureResult.whenComplete { _ in
+                    channel.close(promise: nil)
+                }
+            case .failure(let error):
+                self.state = .end
+                self.delegate.didReceiveError(task: self.task, error)
+                self.promise.fail(error)
+                context.close(promise: nil)
+            }
+        }
+    }
 
-        self.state = .sent
-        self.delegate.didTransmitRequestBody(task: self.task)
+    private func writeBody(request: HTTPClient.Request, context: ChannelHandlerContext) -> EventLoopFuture<Void> {
+        if let body = request.body {
+            return body.provider { part in
+                context.writeAndFlush(self.wrapOutboundOut(HTTPClientRequestPart.body(part)))
```

Comment:
You can use leading-dot syntax here to omit `HTTPClientRequestPart`

Reply:
fixed, thanks!